src/libsyntax/parse/parser.rs

```diff
@@ -233,6 +233,21 @@ macro_rules! maybe_whole {
     )
 }
 
+/// Uses $parse_expr to parse an expression and returns the span of the interpolated
+/// token or the span of the parsed expression, if it was not interpolated
+macro_rules! interpolated_or_expr_span {
```

Comment:
this makes me feel better about the pattern here. (I still worry people may not know when to call this, but at least now it's easy to see and swap in.)

AFAICT, the reason you made this a macro is due to the use of `try!` in the given $parse_expr.

I would prefer you make this a `fn` that returns Result, and shift the trys at the call sites.